### DISCUSSION TITLE
How to inject dependencies when just "sprinkling in" lit components in regular HTML

### DISCUSSION THREAD

**Discussion by sr258** [lit/lit#3539] [Q&A]
Some of my Lit components need to make API calls. I want to decouple the API calls by only using an interface of the API service in the components. That way I can use mocks when I develop them (for quick previews outside of my actual application).

I wonder what the best way of injecting the real API services and mocks is for my setup: At the moment I just treat the lit components like regular HTML elements (I do server side rendering with templates). I basically "sprinkle in" the dynamic lit components while most of the site is static in the browser. That means not my whole application is built with lit but only parts. I can inject data by passing in stringified JSON structures, but I can't do this for the services.

For my mocks I've written a basic JavaScript script that gets all lit elements of a certain type and sets the services manually. Is there a better approach that is more generic?

**Comment by justinfagnani** [comment]
> For my mocks I've written a basic JavaScript script that gets all lit elements of a certain type and sets the services manually

I think this is actually a fine way to do things as a start. If the data you want to set isn't describable in HTML, you'll need some JS glue and this works.

The other two options are to use globals, or to use the Context Protocol.

The reason to use globals would be because it's simple and it's somewhat like the web platform - major API are just available as global objects or functions. The reasons not to use globals are many: you may need different services for different elements, it's relatively unsafe, etc.

Context is a nice approach because it's a way to get data to elements without directly binding to them, with keys, and with the ability to vary the data for different parts of the tree. Normally context is used from JS, with things like [@lit-labs/context](https://www.npmjs.com/package/@lit-labs/context). With this it would't be a _lot_ different from your current approach, except that you don't have to find the elements, they request the data.

Right now `@lit-labs/context` is fairly tied to custom elements because it's vended as reactive controllers that need a host. So you can create your own context-provider custom element and put that at the root of the page.

It might look something like:

```ts
import {ReactiveElement} from 'lit';
import {customElement} from 'lit/decorators.js';
import {provide} from '@lit-labs/context';

export const serviceAContext = createContext<ServiceA>();
export const serviceBContext = createContext<ServiceB>();

@customElement('services-provider')
class ServicesProvider extends ReactiveElement {
  @provide() serviceA = new ServiceA();
  @provide() serviceB = new ServiceB();
}
```

If you wanted to make this more generic and dynamic you could make a context-provider element that has an API for providing a context key/value, like `ProviderElement.provide(context, value)`.
